system/metrics/sinks/prometheus/prometheusSink.hpp

```diff
@@ -0,0 +1,116 @@
+/*################################################################################
+#    HPCC SYSTEMS software Copyright (C) 2021 HPCC Systems®.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+################################################################################
+*/
+
+
+#pragma once
+
+#include "jmetrics.hpp"
+#include "jptree.hpp"
+#include "jstring.hpp"
+#include <thread>
+
+//including cpp-httplib single header file REST client
+//  doesn't work with format-nonliteral as an error
+//
+#if defined(__clang__) || defined(__GNUC__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wformat-nonliteral"
+#endif
+
+#undef INVALID_SOCKET
+#include "httplib.h"
+
+#if defined(__clang__) || defined(__GNUC__)
+#pragma GCC diagnostic pop
+#endif
+
+#undef INVALID_SOCKET
+#include "httplib.h"
+
+#if defined(__clang__) || defined(__GNUC__)
+#pragma GCC diagnostic pop
+#endif
+
+#ifdef _USE_OPENSSL
+#include <openssl/x509v3.h>
+#endif
+
+using namespace hpccMetrics;
+using namespace httplib;
+
+#ifdef PROMETHEUSSINK_EXPORTS
+#define PROMETHEUSSINK_API DECL_EXPORT
+#else
+#define PROMETHEUSSINK_API DECL_IMPORT
+#endif
+
+class PROMETHEUSSINK_API PrometheusMetricSink : public MetricSink
+{
+public:
+    explicit PrometheusMetricSink(const char *name, const IPropertyTree *pSettingsTree);
+    ~PrometheusMetricSink() override = default;
+    static constexpr const char * PROMETHEUS_REPORTER_TYPE = "prometheus";
+    void startServer();
+private:
+    std::thread       m_collectThread;
+    MetricsReporter * m_metricsReporter;
+    StringBuffer      m_metricsSinkName;
+
+    std::atomic<bool>               m_processing{false};
+    int                             m_port;
+    StringBuffer                    m_metricsServiceName;
+    bool                            m_verbose;
+
+    static constexpr const char * PROMETHEUS_METRICS_HTTP_PAGE_TITLE = "HPCC Systems - Prometheus Metrics Service";
+    static constexpr const char * PROMETHEUS_METRICS_SERVICE_RESP_TYPE = "text/html; charset=UTF-8";
+    static constexpr int          DEFAULT_PROMETHEUS_METRICS_SERVICE_PORT = 8767;
+    static constexpr bool         DEFAULT_PROMETHEUS_METRICS_SERVICE_SSL = false;
+    static constexpr const char * DEFAULT_PROMETHEUS_METRICS_SERVICE_NAME = "/metrics";
+    static constexpr const char * HTTPLIB_ERROR_MESSAGE_HEADER_NAME = "EXCEPTION_WHAT";
+    static constexpr const char * PROMETHEUS_METRICS_HTTP_ERROR = "\
```

Comment:
I don't think this is a legal c++ string.  I'm not sure why the compiler is accepting it.  You could either use string concatenation:

const char * s = "abc"
    "def\n"
    "...";

or use the raw string syntax if you want embedded newlines, and no need to quote ".

const char * s = R"!!(<html>
...
)!!";